llvm/lib/Target/AMDGPU/AMDGPUAttributor.cpp

```diff
@@ -168,7 +168,15 @@ class AMDGPUInformationCache : public InformationCache {
     return ST.supportsGetDoorbellID();
   }
 
-  std::pair<unsigned, unsigned> getFlatWorkGroupSizes(const Function &F) {
+  std::optional<std::pair<unsigned, unsigned>>
+  getFlatWorkGroupSizeAttr(const Function &F) const {
+    Attribute A = F.getFnAttribute("amdgpu-flat-work-group-size");
+    if (!A.isStringAttribute())
+      return std::nullopt;
```

Comment:
The presence or absence of an attribute should not impart meaning. A function explicitly marked with the default value should be the same as one with no attribute.

This will also miss applying the calling convention based default range

Reply:
> A function explicitly marked with the default value should be the same as one with no attribute.

I think we still need to honor existing value, no matter whether it is default or not. FWIW, with this change, at least our compiler will not add default value to the attribute. If a default value appears in the attribute, it must be from users.

Reply:
> The presence or absence of an attribute should not impart meaning. A function explicitly marked with the default value should be the same as one with no attribute.

Give what the front end is doing, we will indeed discard the default attribute.